File: week1_1.5/PathCompressQU.cpp
#include "PathCompressQU.h"

PathCompressQU::PathCompressQU(int size){
    id.resize(size);
    for (int i=0;i<size;i++)
        id[i] = i;
}

void PathCompressQU::Union(int a, int b){
    int ra = root(a), rb = root(b);
    id[ra] = rb;
}

bool PathCompressQU::Connected(int a, int b){
    return root(a) == root(b);
}

int PathCompressQU::root(int a){
    while (a != id[a]){
        id[a] = id[id[a]];//compress path, make tree flat
        a = id[a];
    }
    return a;
}

File: src/chapter1_5/PathCompressQU.java
package ch1.DynamicConnection;

import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.Stopwatch;

public class PathCompressQU {
    private int[] id;
    private int[] sz;

    public PathCompressQU(int n) {
        id = new int[n];
        sz = new int[n];
        for (int i = 0; i < n; i++) {
            id[i] = i;
            sz[i] = 1;
        }
    }

    private int find(int a) {
        while (id[a] != a) {
            id[a] = id[id[a]];//Path compress
            a = id[a];
        }
        return a;
    }

    public boolean connected(int a, int b) {
        return find(a) == find(b);
    }

    public void union(int a, int b) {
        int ida = find(a);
        int idb = find(b);
        if (ida == idb) return;
        if (sz[ida] < sz[idb]) {
            id[ida] = idb;
            sz[idb] += sz[ida];
        } else {
            id[idb] = ida;
            sz[ida] += sz[idb];
        }
    }

    public static void main(String[] args) {
        In in = new In(args[0]);
        int n = in.readInt();
        PathCompressQU uf = new PathCompressQU(n);
        Stopwatch watch = new Stopwatch();
        for (int i = 0; i < n; i++) {
            int a = in.readInt();
            int b = in.readInt();
            if (uf.connected(a, b)) {
                ;//StdOut.println("(" + a + ", " + b + ") is connected.");
            } else {
                uf.union(a, b);
            }
        }
        StdOut.println(n + " data: " + watch.elapsedTime() + "s.");
    }
}
